rclcpp/include/rclcpp/publisher_base.hpp

```diff
@@ -99,13 +99,13 @@ class PublisherBase : public std::enable_shared_from_this<PublisherBase>
   /// Get the rcl publisher handle.
   /** \return The rcl publisher handle. */
   RCLCPP_PUBLIC
-  rcl_publisher_t *
+  std::shared_ptr<rcl_publisher_t>
   get_publisher_handle();
 
   /// Get the rcl publisher handle.
   /** \return The rcl publisher handle. */
   RCLCPP_PUBLIC
-  const rcl_publisher_t *
+  const std::shared_ptr<rcl_publisher_t>
```

Comment:
if I'm not getting confused, it should be `std::shared_ptr<const rcl_publisher_t>`

Reply:
This is a copy of the code in https://github.com/ros2/rclcpp/blob/eff11d61bbb60220003779eb4a5a1b205a76da29/rclcpp/include/rclcpp/subscription_base.hpp#L93, but modified for publishers.  As I understand, what this code is doing as-is is making sure that the shared_ptr is const (that is, you can't reassign the returned value).  What you are suggesting is we make the structure const (that is, you can't change the structure).  That's definitely a stronger guarantee, but I don't think that that is the intended use-case.  For instance, one use is https://github.com/ros2/rclcpp/pull/1119/files#diff-8c68c968941e2e05ba8974455ea3f1b5R140 , where we are calling `rcl_return_loaned_message_from_publisher()`.  In turn, the API for that is [here](https://github.com/ros2/rcl/blob/0de31d04830f0b33a3adf4a87e179e0d0d76555b/rcl/include/rcl/publisher.h#L256), where it expects a `const rcl_publisher_t *`.

So while I like the stronger const guarantee in general, I think that is a bigger issue that we'd have to evaluate separately.

Reply:
> This is a copy of the code in https://github.com/ros2/rclcpp/blob/eff11d61bbb60220003779eb4a5a1b205a76da29/rclcpp/include/rclcpp/subscription_base.hpp#L93, but modified for publishers. 

mmm, I didn't check what `Subscription` does (neither why)

> As I understand, what this code is doing as-is is making sure that the shared_ptr is const (that is, you can't reassign the returned value).

Yes

> What you are suggesting is we make the structure const (that is, you can't change the structure).

Yes

> That's definitely a stronger guarantee, but I don't think that that is the intended use-case.  For instance, one use is https://github.com/ros2/rclcpp/pull/1119/files#diff-8c68c968941e2e05ba8974455ea3f1b5R140 , where we are calling `rcl_return_loaned_message_from_publisher()`.  In turn, the API for that is [here](https://github.com/ros2/rcl/blob/0de31d04830f0b33a3adf4a87e179e0d0d76555b/rcl/include/rcl/publisher.h#L256), where it expects a `const rcl_publisher_t *`.
> So while I like the stronger const guarantee in general, I think that is a bigger issue that we'd have to evaluate separately.

* `const rcl_publisher_t *`: The `rcl_publisher_t` structure can't be modified.
* `rcl_publisher_t * const`: The pointer can't be reassigned.

Before, we were returning `const rcl_publisher_t *`.
I don't see why we should change the "constness" when upgrading to a `shared_ptr`.
This is relaxing the previous guarantee.

Reply:
We can't drop the signature; some of the `rcl` functions need a const pointer in order to compile.  But I can change it to what @ivanpauno recommends.

Reply:
What? If you have a non-const pointer, then you can get a const pointer, haha.

Reply:
Hm, you make a good point, but I definitely couldn't compile without this in place.  Let me take a quick look again.

Reply:
> We can't drop the signature; some of the rcl functions need a const pointer in order to compile. But I can change it to what @ivanpauno recommends.

I think there's a bit of confusion.
If the method overload with const qualifier is deleted (with const qualifier I mean the one affecting `this` ptr), then in the places where we pass a `const &` or `const *` to a `rclcpp::Publisher` we cannot call `get_publisher_handle`.

It seems that we're doing that somewhere, and thus that overload can't be removed.

Reply:
Yeah, without this signature, I get the following compile error:

```
In file included from /home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/publisher.hpp:35:0,
                 from /home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/topic_statistics/subscription_topic_statistics.hpp:31,
                 from /home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/subscription.hpp:50,
                 from /home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/any_executable.hpp:25,
                 from /home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/src/rclcpp/any_executable.cpp:15:
/home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/loaned_message.hpp: In constructor ‘rclcpp::LoanedMessage<MessageT, AllocatorT>::LoanedMessage(const rclcpp::PublisherBase&, std::allocator<_Tp1>)’:
/home/ubuntu/ros2_ws/src/ros2/rclcpp/rclcpp/include/rclcpp/loaned_message.hpp:68:35: error: no matching function for call to ‘rclcpp::PublisherBase::get_publisher_handle() const’
         pub_.get_publisher_handle().get(),
```

That comes from this line of code: https://github.com/ros2/rclcpp/blob/1a48a60a75895ae86c1f93a57bc1955b11fb56c9/rclcpp/include/rclcpp/loaned_message.hpp#L68 .  The signature of `rcl_borrow_loaned_message` comes from https://github.com/ros2/rcl/blob/master/rcl/include/rcl/publisher.h#L226 .  I guess it is the lack of `const` on the method itself that is the missing bit, not the const pointer.

Not super important.  Just wanted to followup :).

Reply:
Ah, sorry, yeah we need a const qualified method for this class, it's not that the pointer which is returned is const or not.

Reply:
Well also that signature may be wrong (for LoanedMessage), if it wants to take a const reference to a PublisherBase, but then extract a non-const pointer for the rcl publisher from it. Instead maybe it should take a normal reference or a pointer to a PublisherBase...

Reply:
I've opened https://github.com/ros2/rclcpp/issues/1143 to track this for the future.

Reply:
> Well also that signature may be wrong (for LoanedMessage), if it wants to take a const reference to a PublisherBase, but then extract a non-const pointer for the rcl publisher from it.

`LoanedMessage` is taking a [const rclcpp::PublisherBase & pub](https://github.com/ros2/rclcpp/blob/1a48a60a75895ae86c1f93a57bc1955b11fb56c9/rclcpp/include/rclcpp/loaned_message.hpp#L60), and then extracting a : [**const** rcl_publisher_t * publisher](https://github.com/ros2/rcl/blob/c212a0dba4c0fbb833429e38992a0f765b84910e/rcl/include/rcl/publisher.h#L227).

I don't see where's the issue.
We've already changed the signature to return a `std::shared_ptr<const rcl_publisher_t>` in https://github.com/ros2/rclcpp/pull/1119/commits/b9bf5dc9c29f1972c4bed0a873df98908ee54fd4, so everything looks correct to me.

Reply:
Yeah, ok that's fine then.

Reply:
My (unlikely) concern was that this signature:

https://github.com/ros2/rcl/blob/c212a0dba4c0fbb833429e38992a0f765b84910e/rcl/include/rcl/publisher.h#L227

Is incorrectly accepting a const publisher and later mutating its state anyways, but maybe not, maybe it is just used as a key for lookup in some other structure.

I think this is fine, we don't need to do anything more.